upload.go

```diff
@@ -125,6 +144,118 @@ type s3repo struct {
 	session    *session.Session
 }
 
+func NewB2Repo(opts options) (*b2repo, error) {
+	r := &b2repo{
+		appKey:            opts.B2AppKey,
+		bucket:            opts.B2Bucket,
+		concurrentUploads: opts.B2ConcurrentUploads,
+		ctx:               context.Background(),
+		endpoint:          opts.B2Endpoint,
+		forcePath:         opts.B2ForcePath,
+		keyID:             opts.B2KeyID,
+		region:            opts.B2Region,
+	}
+
+	l.Verbosef("starting b2 client with %d connections to %s %s \n", r.concurrentUploads, r.endpoint, r.bucket)
+	client, err := b2.NewClient(r.ctx, r.keyID, r.appKey)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not create B2 session: %w", err)
+	}
+
+	r.b2Client = client
+
+	bucket, err := r.b2Client.Bucket(r.ctx, r.bucket)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to B2 bucket: %w", err)
+	}
+
+	r.b2Bucket = bucket
+
+	return r, nil
+}
+
+func (r *b2repo) Upload(path string, target string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	w := r.b2Bucket.Object(target).NewWriter(r.ctx)
+	w.ConcurrentUploads = r.concurrentUploads
+
+	l.Infof("uploading %s to B2 bucket %s\n", path, r.bucket)
+	if _, err := io.Copy(w, f); err != nil {
+		w.Close()
+		return err
+	}
+
+	return w.Close()
+}
+
+func (r *b2repo) Download(target string, path string) error {
+	file, err := os.Create(path)
+	if err != nil {
+		return fmt.Errorf("download error: %w", err)
+	}
+	defer file.Close()
+
+	bucket := r.b2Bucket
+
+	l.Infof("downloading %s from B2 bucket %s to %s\n", target, r.bucket, path)
+	remoteFile := bucket.Object(target).NewReader(r.ctx)
+	defer remoteFile.Close()
+
+	localFile, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+
+	if _, err := io.Copy(file, remoteFile); err != nil {
+		localFile.Close()
+		return err
+	}
+	return localFile.Close()
```

Comment:
Here, 2 local files are created and mixed, but only `file` gets data. It seems wrong.

Maybe remove the creation of `file` at the beginning of the function, and use `localFile` only, as the variable name is more readable.

Reply:
Hi,

fixed with including `defer X.close()`, i also did the same pattern in `Upload()`